### PR TITLE
Migrate to ESLint and apply Airbnb JavaScript Style Guide

### PR DESCRIPTION
**Pull request by IGassmann**
This PR migrate from TSLint to ESLint since the former [has been deprecated](https://medium.com/palantir/tslint-in-2019-1a144c2317a9).

It also configures ESLint and Prettier to align themselves with the [Airbnb JavaScript Style Guide](https://airbnb.io/javascript/), which is the most popular style guide. It is well documented and its rules have been thoroughly discussed by the community. Thanks to `typescript-eslint` it can [also be used with TypeScript](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#usage-with-airbnb).

After merging this, the idea is to normalize the linting configuration between the different Nest official projects and documentation as discussed [here](https://github.com/nestjs/nest/issues/2918).

- Closes https://github.com/nestjs/typescript-starter/issues/69.

### CODE REVIEW

File: .prettierrc.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  singleQuote: true,
+  printWidth: 100,
```
Comment (kamilmysliwiec) [inline .prettierrc.js]: Can we remove `printWidth` option here? (and leave the default value)

File: package.json [hunk 2]
```diff
@@ -6,12 +6,12 @@
   "license": "MIT",
   "scripts": {
     "build": "rimraf dist && tsc -p tsconfig.build.json",
-    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
     "start": "ts-node -r tsconfig-paths/register src/main.ts",
     "start:dev": "tsc-watch -p tsconfig.build.json --onSuccess \"node dist/main.js\"",
     "start:debug": "tsc-watch -p tsconfig.build.json --onSuccess \"node --inspect-brk dist/main.js\"",
     "start:prod": "node dist/main.js",
-    "lint": "tslint -p tsconfig.json -c tslint.json",
+    "format": "prettier --write \"**/*.{ts,js,json,md,yml}\" --ignore-path .gitignore --write",
```
Comment (kamilmysliwiec) [inline package.json]: We have `--write` 2 times here

File: src/app.controller.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import { Controller, Get } from '@nestjs/common';
-import { AppService } from './app.service';
+import AppService from './app.service';
 
 @Controller()
-export class AppController {
+export default class AppController {
```
Comment (kamilmysliwiec) [inline src/app.controller.ts]: Default exports are not recommended with Nest and should be permitted rather than suggested

File: src/app.service.ts [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
 
 @Injectable()
-export class AppService {
+export default class AppService {
+
+  private readonly helloWorld = 'Hello World!';
+
   getHello(): string {
-    return 'Hello World!';
+    return this.helloWorld;
```
Comment (kamilmysliwiec) [inline src/app.service.ts]: Can we revert this change to stay consistent with all tutorials & guides & courses on the web?

File: .prettierrc.js (hunk 1 above)
Comment (IGassmann) [reply .prettierrc.js]: The rule being followed here is https://airbnb.io/javascript/#whitespace--max-len.

File: src/app.controller.ts (hunk 3 above)
Comment (IGassmann) [reply src/app.controller.ts]: The rule being followed here is https://airbnb.io/javascript/#modules--prefer-default-export.

**Comment by IGassmann** [comment]
Concerning the changes in the code style. I've adjusted the code to follow the rules of the Airbnb Style Guide.

When adopting a style guide, I highly recommend not tweaking the rules of the guide to fit personal preferences. The whole point of adopting a guide is consistency and to get rid of bikeshedding.

Many organizations and projects use a style guide and are intransigent with their specified rules. Google is an example of such organization:

> On matters of style, the style guide is the absolute authority. Any purely style point (whitespace, etc.) that is not in the style guide is a matter of personal preference. The style should be consistent with what is there. If there is no previous style, accept the author’s.

Source: https://google.github.io/eng-practices/review/reviewer/standard.html

File: .prettierrc.js (hunk 1 above)
Comment (kamilmysliwiec) [reply .prettierrc.js]: Yeah, but at the same time, 120 is the default option proposed by prettier. So far, we've used prettier's one so I think that we shouldn't change this.

**Comment by kamilmysliwiec** [comment]
@IGassmann I haven't proposed changes that are in line with my preferences, but the framework recommendations according to the documentation and the entire ecosystem. `default` shouldn't be used.

File: src/main.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { NestFactory } from '@nestjs/core';
-import { AppModule } from './app.module';
+import AppModule from './app.module';
 
-async function bootstrap() {
+async function bootstrap(): Promise<void> {
```
Comment (kamilmysliwiec) [inline src/main.ts]: `Promise<void>` doesn't bring any relevant & useful information here. I know that it's forced by the Style Guide, but I doubt we should follow this.

**Comment by IGassmann** [comment]
> @IGassmann I haven't proposed changes that are in line with my preferences, but the framework recommendations according to the documentation and the entire ecosystem. `default` shouldn't be used.

Yes. I should have written preferences in general.

The point I'm trying to communicate is the value of having a documented style guide that we can use as a reference and not break its rules. That would mean updating the style of the projects and documentation to align with the style of the guide.

Another proposition I would have if we want to keep some existing styles is to write a Nest Style Guide and build an ESLint plugin or adopt another style guide that is closer to Nest's style and also has the tools ready. Maybe [Google's one ](https://google.github.io/styleguide/jsguide.html) would be a viable solution. It is adopted by the Angular projects.

**Comment by kamilmysliwiec** [comment]
It doesn't make too much sense to create a dedicated style guide if we just want to override a few rules to make all repos & documentation consistent. I do agree with the majority of AirBnb rules, although, some decisions aren't the best and we shouldn't just stick to it because the inherited style guide enforces them. Both eslint and tslint allow to override rules and that's what makes them powerful. Obviously, we should reduce the number of them **as much as possible**, but people eventually will use either our CLI, typescript-starter, samples from the main repo or docs to start their projects. If we ensure that all these repos share the same eslint file with the same set of rules - that would be totally fine, regardless of a couple of modified rules.

**Comment by IGassmann** [comment]
@BrunnerLivio @kamilmysliwiec makes sense for me to wait for Angular's migration. They will probably set it up with the [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html) rules, which is also a well-documented style guide.

**Comment by pumano** [comment]
@IGassmann you going to good direction, but airbnb use react where `export default` is preferred. Angular not use default exports. You should explicitly import what you need. Its reason, why that rule not fit here. NestJS currently going tightly with Angular.

**Comment by IGassmann** [comment]
@pumano yes, I could see how Google's style guide would be a better fit if the Nest wants to align with the Angular project.

However, Google's style guide has some questionable decisions too like requering JSDoc for all classes, methods, and functions. Other things are also different from what Nest has been doing, like asking for annotation to be written in camelCase.

One could also argue that Nest is agnostic of frontend frameworks, and then there's no necessity for it to be aligned with Angular style.

File: src/app.service.ts (hunk 4 above)
Comment (rafalpetryka) [reply src/app.service.ts]: So you would like to turn off this eslint rule?
https://eslint.org/docs/rules/class-methods-use-this

**Comment by THPubs** [comment]
Hey guys, shall we finalize this? If there's a debate in which one to use, Google or Airbnb, you can refer to this article: https://medium.com/@uistephen/style-guides-for-linting-ecmascript-2015-eslint-common-google-airbnb-6c25fd3dff0 or some others around there. I personally go with Airbnb since it's the most popular one and also since it have more React support. But it's up-to you.

If @kamilmysliwiec want to modify it, I'm completely ok with it since it does look like this framework need some modifications. But the thing is, tslint have announce their deprecation February this year. Now we have waited too long for it's adoption which is not good at all. We can't wait for Angular guys. So, shall we come to any common agreement and move forward with this? @IGassmann If you need any help I'm free. 🙂  Will it be ok if I help with the changes?

**Comment by IGassmann** [comment]
I believe it's still okay for us to wait. TSLint is still [accepting PRs for bug and security fixes](https://github.com/palantir/tslint/issues/4534). However, I also understand the need to get this done, so new NestJS projects won't need to make the switch to ESLint themselves.

The angular project [might be preparing](https://github.com/angular/angular-cli/issues/13732#issuecomment-503722431) an ESLint configuration that would accommodate [their style guide](https://angular.io/guide/styleguide), which ain't exactly the Google one. We have adopted internally the Airbnb Style Guide since September for our NestJS project. It's been working fine, even though there are some divergences with the Angular rules, such as the use of default exports.

**Comment by THPubs** [comment]
@IGassmann No offense but this is not an Angular project. This is a back-end framework used by many other developers as well. I agree this project follow many good standards set by Angular. But that does not mean we need to follow them in all the other cases. Also note that Angular focus on Front end and this focus on the Back end.

**Comment by IGassmann** [comment]
@THPubs no offense taken. I'm bringing Angular, because the project was inspired by it and many have come from it. I personally don't use Angular.

In my own opinion, I would adopt Airbnb, or create a custom one, but have it documented somewhere by forking the documentation of an existing one.

**Comment by kamilmysliwiec** [comment]
@THPubs there is a good chance we won't follow exactly the same set of rules as Angular. However, since they already started working on this, it would be great to ensure our ESLint configurations are as close as possible to each other. Let's be patient, they should release it shortly.

**Comment by trylovetom** [comment]
Any update bro? It's 2020 now.

**Comment by bukhalo** [comment]
Switching to ESLint is a forced measure because TSLint will not be updated soon (sadly, ESLint is much slower). But Airbnb rules are a bad idea, I think. It has a lot of front-end specific rules and it doesn't let you work quietly. Instead of focusing on writing code, you have to worry about satisfying Airbnb rules. Can we make our own rules for Nest based on Airbnb experience?

**Comment by zbeyens** [comment]
I would extend airbnb as it nearly became a convention to do it. Then I just disabled the unsatisfying rules. I am now using the same eslint config for my front-end and back-end without any conflict.

**Comment by IGassmann** [comment]
Extend/Fork Airbnb's would be a good strategy. I would also add TypeScript specific rules since Airbnb targets only JavaScript. Nest would then have its own set of rules and style guide like other frameworks out there.

**Comment by kamilmysliwiec** [comment]
Let's track this here https://github.com/nestjs/typescript-starter/pull/161

Initially, we wanted to wait until Angular publishes a new release. Unfortunately, Angular 9 (new major release) is already a few months delayed, and the migration (from tslint to eslint) has been postponed until Angular 10. 

More information in the PR.
